kmod/src/xattr.c

```diff
@@ -498,6 +518,52 @@ static int parse_totl_u64(const char *s, int len, u64 *res)
 	return kstrtoull(str, 0, res) != 0 ? -EINVAL : 0;
 }
 
+static int parse_totl_u32(const char *s, int len, unsigned long int *res)
+{
+	char str[SCOUTFS_XATTR_MAX_TOTL_U32 + 1];
+
+	if (len <= 0 || len >= ARRAY_SIZE(str) || s[0] == '+' || s[len - 1] == '\n')
+		return -EINVAL;
+
+	memcpy(str, s, len);
+	str[len] = '\0';
+
+	return kstrtoul(str, 0, res) != 0 ? -EINVAL : 0;
+}
+
+static int parse_worm_timeval(struct scoutfs_inode_info *si,
+				const char *name, int name_len)
+{
+	u64 tv_sec;
+	unsigned long tv_nsec;
+	int start = 0;
+	int len;
+	int ret = 0;
+	int i;
+
+	for (i = 0; i < name_len; i++) {
+		if (name[i] != '.')
+			continue;
+
+		ret = parse_totl_u64(&name[start], i, &tv_sec);
+		if (ret < 0)
+			goto out;
+		len = name_len-i-1;
+		ret = parse_totl_u32(&name[i+1], len, &tv_nsec);
+		if (ret < 0)
+			goto out;
+	}
+
+	if (ret == 0) {
```

Comment:
We won't set the inode fields here as the xattrs parsed 'cause we don't know if the xattr_set has succeeded yet or not.   That's why we parse into an on-stack timespec and only assign it to the si fields (or clear them) at the very end of xattr_set if it succeeded before we call _update_inode_item.